# Data Cleaning and Organization

> - The data used is available at: http://football-data.co.uk/englandm.php
>
> - The used data starts from season 2000-2001 to 2020-2021. 

## Importing libs.

In [97]:
import numpy as np
import pandas as pd
from datetime import datetime as dt
import itertools

# Reading data file.

In [98]:
location = "Datasets/"

raw_data_1 = pd.read_csv(location + '00-01.csv')
raw_data_2 = pd.read_csv(location + '01-02.csv')
raw_data_3 = pd.read_csv(location + '02-03.csv')
raw_data_4 = pd.read_csv(location + '03-04.csv')
raw_data_5 = pd.read_csv(location + '04-05.csv')
raw_data_6 = pd.read_csv(location + '05-06.csv')
raw_data_7 = pd.read_csv(location + '06-07.csv')
raw_data_8 = pd.read_csv(location + '07-08.csv')
raw_data_9 = pd.read_csv(location + '08-09.csv')
raw_data_10 = pd.read_csv(location + '09-10.csv')
raw_data_11 = pd.read_csv(location + '10-11.csv')
raw_data_12 = pd.read_csv(location + '11-12.csv')
raw_data_13 = pd.read_csv(location + '12-13.csv')
raw_data_14 = pd.read_csv(location + '13-14.csv')
raw_data_15 = pd.read_csv(location + '14-15.csv')
raw_data_16 = pd.read_csv(location + '15-16.csv')
raw_data_17 = pd.read_csv(location + '16-17.csv')
raw_data_18 = pd.read_csv(location + '17-18.csv')
raw_data_19 = pd.read_csv(location + '19-20.csv')
raw_data_20 = pd.read_csv(location + '20-21.csv')

In [99]:
raw_data_3.head()

Div        Date   HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  ...  \
0  E0  17/08/2002  Blackburn  Sunderland     0     0   D     0     0   D  ...   
1  E0  17/08/2002   Charlton     Chelsea     2     3   A     2     1   H  ...   
2  E0  17/08/2002    Everton   Tottenham     2     2   D     1     0   H  ...   
3  E0  17/08/2002     Fulham      Bolton     4     1   H     3     1   H  ...   
4  E0  17/08/2002      Leeds    Man City     3     0   H     2     0   H  ...   

  HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  
0   5    3  14  11   9   1   1   2   0   0  
1   5   12  10  12   3   6   0   3   1   0  
2   9    5  18   4  10   5   1   1   0   0  
3   6    1  16  12   7   4   1   2   0   0  
4   8   10  13  13   2   7   1   1   0   0  

[5 rows x 23 columns]

In [100]:
raw_data_17.head()

Div        Date        HomeTeam      AwayTeam  FTHG  FTAG FTR  HTHG  HTAG  \
0  E0  11/08/2017         Arsenal     Leicester     4     3   H     2     2   
1  E0  12/08/2017        Brighton      Man City     0     2   A     0     0   
2  E0  12/08/2017         Chelsea       Burnley     2     3   A     0     3   
3  E0  12/08/2017  Crystal Palace  Huddersfield     0     3   A     0     2   
4  E0  12/08/2017         Everton         Stoke     1     0   H     1     0   

  HTR  ... BbAv<2.5  BbAH  BbAHh  BbMxAHH  BbAvAHH  BbMxAHA  BbAvAHA   PSCH  \
0   D  ...     2.32    21  -1.00     1.91     1.85     2.10     2.02   1.49   
1   D  ...     2.27    20   1.50     1.95     1.91     2.01     1.96  11.75   
2   A  ...     2.23    20  -1.75     2.03     1.97     1.95     1.90   1.33   
3   A  ...     1.72    18  -0.75     2.10     2.05     1.86     1.83   1.79   
4   H  ...     1.76    19  -0.75     1.94     1.90     2.01     1.98   1.82   

   PSCD   PSCA  
0  4.73   7.25  
1  6.15   1.29  
2  5.40  12.25  
3  3.56   5.51  
4  3.49   5.42  

[5 rows x 65 columns]

In [101]:
raw_data_1.shape

(380, 28)

All data sets has 380 entry. this represents the total number of matches per season.

In [102]:
raw_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 380 entries, 0 to 379
Data columns (total 28 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Div         380 non-null    object
 1   Date        380 non-null    object
 2   HomeTeam    380 non-null    object
 3   AwayTeam    380 non-null    object
 4   FTHG        380 non-null    int64 
 5   FTAG        380 non-null    int64 
 6   FTR         380 non-null    object
 7   HTHG        380 non-null    int64 
 8   HTAG        380 non-null    int64 
 9   HTR         380 non-null    object
 10  Attendance  380 non-null    int64 
 11  Referee     380 non-null    object
 12  HS          380 non-null    int64 
 13  AS          380 non-null    int64 
 14  HST         380 non-null    int64 
 15  AST         380 non-null    int64 
 16  HHW         380 non-null    int64 
 17  AHW         380 non-null    int64 
 18  HC          380 non-null    int64 
 19  AC          380 non-null    int64 
 20  HF        

Date should be transformed from object tot date-time.

## Date correction

In [103]:
raw_data_1['Date'] = pd.to_datetime(raw_data_1['Date'])
raw_data_2['Date'] = pd.to_datetime(raw_data_2['Date'])
raw_data_3['Date'] = pd.to_datetime(raw_data_3['Date'])
raw_data_4['Date'] = pd.to_datetime(raw_data_4['Date'])
raw_data_5['Date'] = pd.to_datetime(raw_data_5['Date'])
raw_data_6['Date'] = pd.to_datetime(raw_data_6['Date'])
raw_data_7['Date'] = pd.to_datetime(raw_data_7['Date'])
raw_data_8['Date'] = pd.to_datetime(raw_data_8['Date'])
raw_data_9['Date'] = pd.to_datetime(raw_data_9['Date'])
raw_data_10['Date'] = pd.to_datetime(raw_data_10['Date'])
raw_data_11['Date'] = pd.to_datetime(raw_data_11['Date'])
raw_data_12['Date'] = pd.to_datetime(raw_data_12['Date'])
raw_data_13['Date'] = pd.to_datetime(raw_data_13['Date'])
raw_data_14['Date'] = pd.to_datetime(raw_data_14['Date'])
raw_data_15['Date'] = pd.to_datetime(raw_data_15['Date'])
raw_data_16['Date'] = pd.to_datetime(raw_data_16['Date'])
raw_data_17['Date'] = pd.to_datetime(raw_data_17['Date'])
raw_data_18['Date'] = pd.to_datetime(raw_data_18['Date'])
raw_data_19['Date'] = pd.to_datetime(raw_data_19['Date'])
raw_data_20['Date'] = pd.to_datetime(raw_data_20['Date'])

### Check! 

In [104]:
raw_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 380 entries, 0 to 379
Data columns (total 28 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Div         380 non-null    object        
 1   Date        380 non-null    datetime64[ns]
 2   HomeTeam    380 non-null    object        
 3   AwayTeam    380 non-null    object        
 4   FTHG        380 non-null    int64         
 5   FTAG        380 non-null    int64         
 6   FTR         380 non-null    object        
 7   HTHG        380 non-null    int64         
 8   HTAG        380 non-null    int64         
 9   HTR         380 non-null    object        
 10  Attendance  380 non-null    int64         
 11  Referee     380 non-null    object        
 12  HS          380 non-null    int64         
 13  AS          380 non-null    int64         
 14  HST         380 non-null    int64         
 15  AST         380 non-null    int64         
 16  HHW         380 non-null  

:heavy_minus_sign:


## Merge data together

In [105]:
total_df = pd.concat([raw_data_1,
          raw_data_2,
          raw_data_3,
          raw_data_4,
          raw_data_5,
          raw_data_6,
          raw_data_7,
          raw_data_8,
          raw_data_9,
          raw_data_10,
          raw_data_11,
          raw_data_12,
          raw_data_13,
          raw_data_14,
          raw_data_15,
          raw_data_16,
          raw_data_17,
          raw_data_18,
          raw_data_19,
          raw_data_20], ignore_index=True)

In [106]:
total_df = total_df.drop(labels=total_df.columns[24:144], axis=1)
total_df = total_df.drop(labels=['Div'], axis=1)

In [107]:
total_df.columns

Index(['Date', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG', 'FTR', 'HTHG', 'HTAG',
       'HTR', 'Attendance', 'Referee', 'HS', 'AS', 'HST', 'AST', 'HHW', 'AHW',
       'HC', 'AC', 'HF', 'AF', 'HO', 'AO'],
      dtype='object')

In [108]:
total_df.head()

Date  HomeTeam       AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0 2000-08-19  Charlton       Man City     4     0   H     2     0   H   
1 2000-08-19   Chelsea       West Ham     4     2   H     1     0   H   
2 2000-08-19  Coventry  Middlesbrough     1     3   A     1     1   D   
3 2000-08-19     Derby    Southampton     2     2   D     1     2   A   
4 2000-08-19     Leeds        Everton     2     0   H     2     0   H   

   Attendance  ... HST  AST  HHW  AHW  HC  AC  HF  AF   HO   AO  
0     20043.0  ...  14    4  2.0  1.0   6   6  13  12  8.0  6.0  
1     34914.0  ...  10    5  1.0  0.0   7   7  19  14  2.0  3.0  
2     20624.0  ...   3    9  0.0  1.0   8   4  15  21  1.0  3.0  
3     27223.0  ...   4    6  0.0  0.0   5   8  11  13  0.0  2.0  
4     40010.0  ...   8    6  0.0  0.0   6   4  21  20  6.0  1.0  

[5 rows x 23 columns]

# Exporting the entire data set

In [109]:
loc = "Datasets/"
total_df.to_csv(loc + 'Total dataset file.csv')